cmd/webhook/main.go

```diff
@@ -18,135 +18,122 @@ package main
-	ctxFunc := func(ctx context.Context) context.Context {
-		// TODO: implement upgrades when eventing needs it:
-		//  return v1beta1.WithUpgradeViaDefaulting(store.ToContext(ctx))
-		return ctx
-	}
+		// Name of the validation webhook.
+		"validation.webhook.events.cloud.google.com",
 
-	resourceAdmissionController := webhook.NewResourceAdmissionController(resourceHandlers, options, true)
-	admissionControllers := map[string]webhook.AdmissionController{
-		options.ResourceAdmissionControllerPath: resourceAdmissionController,
-	}
+		// The path on which to serve the webhook.
+		"/validation",
 
-	controller, err := webhook.New(kubeClient, options, admissionControllers, logger.Sugar(), ctxFunc)
+		// The resources to validate and default.
+		types,
 
-	if err != nil {
-		logger.Fatal("Failed to create admission controller", zap.Error(err))
-	}
+		// A function that infuses the context passed to Validate/SetDefaults with custom metadata.
+		func(ctx context.Context) context.Context {
+			// return v1.WithUpgradeViaDefaulting(store.ToContext(ctx))
+			return ctx
+		},
 
-	if err = controller.Run(ctx.Done()); err != nil {
-		logger.Fatal("Failed to start the admission controller", zap.Error(err))
-	}
+		// Whether to disallow unknown fields.
+		true,
+	)
 }
 
-func main() {
-	handlers := map[schema.GroupVersionKind]webhook.GenericCRD{
-		messagingv1alpha1.SchemeGroupVersion.WithKind("Channel"):       &messagingv1alpha1.Channel{},
-		messagingv1alpha1.SchemeGroupVersion.WithKind("Decorator"):     &messagingv1alpha1.Decorator{},
-		eventsv1alpha1.SchemeGroupVersion.WithKind("Storage"):          &eventsv1alpha1.Storage{},
-		eventsv1alpha1.SchemeGroupVersion.WithKind("Scheduler"):        &eventsv1alpha1.Scheduler{},
-		eventsv1alpha1.SchemeGroupVersion.WithKind("PubSub"):           &eventsv1alpha1.PubSub{},
-		pubsubv1alpha1.SchemeGroupVersion.WithKind("PullSubscription"): &pubsubv1alpha1.PullSubscription{},
-		pubsubv1alpha1.SchemeGroupVersion.WithKind("Topic"):            &pubsubv1alpha1.Topic{},
-	}
-	SharedMain(handlers)
+func NewConfigValidationController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	return configmaps.NewAdmissionController(ctx,
+
+		// Name of the configmap webhook.
+		"config.webhook.events.cloud.google.com",
+
+		// The path on which to serve the webhook.
+		"/config-validation",
+
+		// The configmaps to validate.
+		configmap.Constructors{
+			tracingconfig.ConfigName: tracingconfig.NewTracingConfigFromConfigMap,
+			// metrics.ConfigMapName():   metricsconfig.NewObservabilityConfigFromConfigMap,
+			logging.ConfigMapName(): logging.NewConfigFromConfigMap,
+		},
+	)
 }
 
-func flush(logger *zap.Logger) {
-	_ = logger.Sync()
-	metrics.FlushExporter()
+func main() {
+	// Set up a signal context with our webhook options
+	ctx := webhook.WithOptions(signals.NewContext(), webhook.Options{
+		ServiceName: logconfig.WebhookName(),
+		Port:        8443,
+		SecretName:  "webhook-certs",
```

Comment:
This needs a comment explaining that this value must match the name of the Secret created in config.

Reply:
Done!